File: data_validation_with_score.scala
import java.sql.DriverManager
import java.sql.Connection

object DataQualityValidation {
  def main(args: Array[String]): Unit = {
    // Colunas e tabelas que serão validadas
    val columnsToValidate = List("CustomerName", "Address", "PhoneNumber")
    val tablesToValidate = List("Customers")

    // Criar conexão com o banco de dados
    val driver = "org.postgresql.Driver"
    val url = "jdbc:postgresql://localhost:5432/mydatabase"
    val username = "username"
    val password = "password"
    var connection:Connection = null
    Class.forName(driver)
    connection = DriverManager.getConnection(url, username, password)

    // Criar tabela para armazenar os resultados da validação
    val statement = connection.createStatement()
    val createTableQuery = "CREATE TABLE DataQualityValidation (ColumnName VARCHAR(100), InvalidCount INT, TotalCount INT, Score DOUBLE PRECISION)"
    statement.execute(createTableQuery)

    // Percorrer as colunas e tabelas para validação
    for (columnToValidate <- columnsToValidate) {
      for (tableToValidate <- tablesToValidate) {
        // Realizar validação de qualidade de dados
        val selectQuery = s"SELECT COUNT(*) AS TotalCount, COUNT(${columnToValidate}) AS ValidCount FROM ${tableToValidate} WHERE ${columnToValidate} IS NOT NULL AND ${columnToValidate} != ''"
        val resultSet = statement.executeQuery(selectQuery)
        resultSet.next()
        val totalCount = resultSet.getInt("TotalCount")
        val validCount = resultSet.getInt("ValidCount")
        val invalidCount = totalCount - validCount
        val score = 1 - (invalidCount.toDouble / totalCount.toDouble)

        // Inserir resultados da validação na tabela
        val insertQuery = s"INSERT INTO DataQualityValidation (ColumnName, InvalidCount, TotalCount, Score) VALUES ('${columnToValidate}', ${invalidCount}, ${totalCount}, ${score})"
        statement.execute(insertQuery)
      }
    }

    // Exibir resultados da validação
    val selectQuery = "SELECT * FROM DataQualityValidation"
    val resultSet = statement.executeQuery(selectQuery)
    while (resultSet.next()) {
      val columnName = resultSet.getString("ColumnName")
      val invalidCount = resultSet.getInt("InvalidCount")
      val totalCount = resultSet.getInt("TotalCount")
      val score = resultSet.getDouble("Score")
      println(s"${columnName}: InvalidCount=${invalidCount}, TotalCount=${totalCount}, Score=${score}")
    }

    // Fechar conexão e limpar tabela
    statement.execute("DROP TABLE DataQualityValidation")
    connection.close()
  }
}

File: data_validation_with_score.py
import pandas as pd

# Colunas e tabelas que serão validadas
columns_to_validate = ['CustomerName', 'Address', 'PhoneNumber']
tables_to_validate = ['Customers']

# Criar dataframe para armazenar os resultados da validação
data_quality_validation = pd.DataFrame(columns=['ColumnName', 'InvalidCount', 'TotalCount', 'Score'])

# Percorrer as colunas e tabelas para validação
for column_to_validate in columns_to_validate:
    for table_to_validate in tables_to_validate:
        # Realizar validação de qualidade de dados
        df = pd.read_sql(f"SELECT COUNT(*) AS TotalCount, COUNT({column_to_validate}) AS ValidCount FROM {table_to_validate} WHERE {column_to_validate} IS NOT NULL AND {column_to_validate} != ''", connection)
        invalid_count = df['TotalCount'] - df['ValidCount']
        score = 1 - (invalid_count / df['TotalCount'])
        
        # Armazenar resultados da validação no dataframe
        data_quality_validation = data_quality_validation.append({
            'ColumnName': column_to_validate,
            'InvalidCount': invalid_count,
            'TotalCount': df['TotalCount'],
            'Score': score
        }, ignore_index=True)

# Exibir resultados da validação
print(data_quality_validation)

# Limpar dataframe
data_quality_validation = data_quality_validation.iloc[0:0]

